### PR TITLE
Fix borg upgrade, issue #466

### PR DESCRIPTION
**Pull request by ThomasWaldmann**

### CODE REVIEW

File: borg/upgrader.py [hunk 1]
```diff
@@ -33,9 +37,19 @@ def upgrade(self, dryrun=True, inplace=False):
...
                 shutil.copytree(self.path, backup, copy_function=os.link)
+                # we need to create a real copy (not hardlink copy) of index.* and hints.*
+                # otherwise the backup copy will get modified.
+                transaction_id = self.get_index_transaction_id()
+                for name in ['index', 'hints']:
+                    fname = "%s.%d" % (name, transaction_id)
+                    fname_orig = os.path.join(self.path, fname)
+                    fname_backup = os.path.join(backup, fname)
+                    os.remove(fname_backup)
+                    shutil.copy(fname_orig, fname_backup)
```
Comment (anarcat) [inline borg/upgrader.py]: i believe this is incorrect: the problem was that we were doing inplace header changes on the index files later, not that we were not copying them. this introduces an inconsistency on how segments are treated vs indexes and cache...

File: borg/upgrader.py [hunk 2]
```diff
@@ -48,13 +62,21 @@ def upgrade(self, dryrun=True, inplace=False):
         self.lock = UpgradableLock(os.path.join(self.path, 'lock'),
                                    exclusive=True).acquire()
         try:
+            self.convert_repo_index(dryrun)
             self.convert_cache(dryrun)
             self.convert_segments(segments, dryrun=dryrun, inplace=inplace)
```
Comment (anarcat) [inline borg/upgrader.py]: instead, inplace= should be passed here as well, i think.

**Comment by anarcat** [comment]
also, this lacks tests for the bugs it solves.

File: borg/upgrader.py (hunk 1 above)
Comment (ThomasWaldmann) [inline borg/upgrader.py]: your header changing code works for the segment files, so that is not the problem.
i suspect that the open/close modifies the index.xx files before that happens.
btw, you did the open() explicitly, but it happens implicitly in Repository.__init__.

Comment (anarcat) [inline borg/upgrader.py]: the same code is used for segments replacement than for index and cache files, namely `header_replace()`. for the segment files, the `inplace` argument is passed. for the others, no. that was the bug, and the proper fix.

as for `open()` vs `__init__()` you are probably right, but that seems unrelated...

Comment (ThomasWaldmann) [inline borg/upgrader.py]: ok, I am checking this right now, whether it solves the issue.

File: borg/upgrader.py (hunk 2 above)
Comment (ThomasWaldmann) [inline borg/upgrader.py]: done, see latest commit

Comment (anarcat) [inline borg/upgrader.py]: cool, that's what i meant. :)
